presto-hive/src/main/java/io/prestosql/plugin/hive/rubix/RubixInitializer.java

```diff
@@ -64,17 +76,13 @@ public void initializeRubix(NodeManager nodeManager)
     {
         ExecutorService initializerService = Executors.newSingleThreadExecutor();
         ListenableFuture<?> nodeJoinFuture = MoreExecutors.listeningDecorator(initializerService).submit(() ->
-        {
-            while (!(nodeManager.getAllNodes().contains(nodeManager.getCurrentNode()) &&
-                    nodeManager.getAllNodes().stream().anyMatch(Node::isCoordinator))) {
-                try {
-                    Thread.sleep(100);
-                }
-                catch (InterruptedException e) {
-                    throw new RuntimeException(e);
-                }
-            }
-        });
+                Failsafe.with(RETRY_POLICY).run(() -> {
+                    if (!nodeManager.getAllNodes().contains(nodeManager.getCurrentNode()) ||
+                            !nodeManager.getAllNodes().stream().anyMatch(Node::isCoordinator)) {
+                        // Failsafe will propagate this exception only when timeout reached.
```

Comment:
> Why it's not readable? We use something similar in: io/prestosql/tests/product/launcher/testcontainers/SelectedPortWaitStrategy.java:70. This approach also provides nice exception message

The fact that you needed to put comment means it is not readable. We are throwing "timout" exception ever loop but retry.
I would suggest to use Failesafe's timeout mechanism like described in https://jodah.net/failsafe/execution-cancellation/ . And then catch `TimeoutExceededException` and wrap it to `PrestoException`

Reply:
@findepi what do you think?

Reply:
@sopel39 i wish i could help. I would need to understand the mechanics better though. (https://github.com/prestosql/presto/pull/3749#discussion_r426147252)